### PR TITLE
8283368: G1: Remove G1SegmentedArraySegment MEMFLAGS template parameter

### PR DESCRIPTION
**Pull request by walulyai**
Hi all,

Please review this change to remove the MEMFLAGS template parameter for G1SegmentedArraySegment and make it a runtime configuration parameter. This reduces on the code bloat due to templating because many of the methods in G1SegmentedArraySegment are independent of the template parameter.

To achieve this we added factory methods for handling allocation and deallocation. Having the factory methods allows us to embed the array data with the G1SegmentedArraySegment struct thus limiting the allocation to a single allocation as well as the deallocation. One upside to this approach is that we limit the allocation and deallocation overheads. 

Thanks
Ivan

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283368](https://bugs.openjdk.java.net/browse/JDK-8283368): G1: Remove G1SegmentedArraySegment MEMFLAGS template parameter

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to ea77e81b2fe412447cf6a1153cf42e9835658c8f

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7891/head:pull/7891` \
`$ git checkout pull/7891`

Update a local copy of the PR: \
`$ git checkout pull/7891` \
`$ git pull https://git.openjdk.java.net/jdk pull/7891/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7891`

View PR using the GUI difftool: \
`$ git pr show -t 7891`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7891.diff">https://git.openjdk.java.net/jdk/pull/7891.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — APPROVED** [review]
Lgtm. Would have been nice to have done the refactoring and the segment inlining separately.

File: src/hotspot/share/gc/g1/g1SegmentedArray.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+G1SegmentedArraySegment* G1SegmentedArray::create_new_segment(G1SegmentedArraySegment* const prev) {
+  // Take an existing segment if available.
+  G1SegmentedArraySegment* next = _free_segment_list->get();
+  if (next == nullptr) {
+    uint prev_num_slots = (prev != nullptr) ? prev->num_slots() : 0;
+    uint num_slots = _alloc_options->next_num_slots(prev_num_slots);
+
+    next = G1SegmentedArraySegment::create_segment(slot_size(), num_slots, prev, _alloc_options->mem_flag());
+    // next = new G1SegmentedArraySegment(slot_size(), num_slots, prev);
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1SegmentedArray.cpp:107]: Seems unneeded.

File: src/hotspot/share/gc/g1/g1SegmentedArray.hpp [hunk 2]
```diff
@@ -92,12 +109,11 @@ class G1SegmentedArraySegment : public CHeapObj<flag> {
...
 // performed by multiple threads at the same time.
 // Counts and memory usage are current on a best-effort basis if accessed concurrently.
-template<MEMFLAGS flag>
 class G1SegmentedArrayFreeList {
-  static G1SegmentedArraySegment<flag>* volatile* next_ptr(G1SegmentedArraySegment<flag>& segment) {
+  static G1SegmentedArraySegment* volatile* next_ptr(G1SegmentedArraySegment& segment) {
     return segment.next_addr();
   }
-  typedef LockFreeStack<G1SegmentedArraySegment<flag>, &G1SegmentedArrayFreeList::next_ptr> SegmentStack;
+  typedef LockFreeStack<G1SegmentedArraySegment, &G1SegmentedArrayFreeList::next_ptr> SegmentStack;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1SegmentedArray.hpp:116]: `using` to be consistent?

**Review by albertnetymk — APPROVED** [review]

**Review by tschatzl — APPROVED** [review]

**Comment by walulyai** [comment]
Thanks @albertnetymk and @tschatzl  for the reviews.

/integrate

**Pull request closed without merging**